Binary and Scalar Embedding Quantization for Significantly Faster & Cheaper Retrieval (#1934)
huggingface/blog: embedding-quantization.md

        - [Scalar Quantization in Sentence Transformers](#scalar-quantization-in-sentence-transformers)
        - [Scalar Quantization in Vector Databases](#scalar-quantization-in-vector-databases)
    + [Combining Binary and Scalar Quantization](#combining-binary-and-scalar-quantization)
    + [Quantization Experiments](#quantization-experiments)
    + [Influence of Rescoring](#influence-of-rescoring)
        - [Binary Rescoring](#binary-rescoring)
        - [Scalar (Int8) Rescoring](#scalar-int8-rescoring)
        - [Retrieval Speed](#retrieval-speed)
    + [Performance Summarization](#performance-summarization)
    + [Demo](#demo)
    + [Try it yourself](#try-it-yourself)
    + [Future work:](#future-work)
    + [Acknowledgments](#acknowledgments)
    + [Citation](#citation)
    + [References](#references)

## Why Embeddings?

Embeddings are one of the most versatile tools in natural language processing, supporting a wide variety of settings and use cases. In essence, embeddings are numerical representations of more complex objects, like text, images, audio, etc. Specifically, the objects are represented as n-dimensional vectors. 

> After transforming the complex objects, you can then determine their similarity by calculating the similarity of the respective embeddings! This is crucial for a massive amount of use cases: it serves as the backbone for recommendation systems, retrieval, one-shot or few-shot learning, outlier detection, similarity search, paraphrase detection, clustering, classification, and much more.

Reviewer: ```suggestion
After transforming the complex objects, you can determine their similarity by calculating the similarity of the respective embeddings! This is crucial for many use cases: it serves as the backbone for recommendation systems, retrieval, one-shot or few-shot learning, outlier detection, similarity search, paraphrase detection, clustering, classification, and much more.
```

| 3072                | [OpenAI text-embedding-3-large](https://openai.com/blog/new-embedding-models-and-api-updates) |1144.40GB<br>$4,348 / mo|2861.02GB<br>$10,871 / mo|11444.09GB<br>$43,487 / mo|

## Improving scalability

There are several ways to approach the challenges of scaling embeddings. The most common approach is dimensionality reduction, such as [PCA](https://en.wikipedia.org/wiki/Principal_component_analysis). However, classic dimensionality reduction -- like PCA methods -- [tends to perform poorly when used with embeddings](https://arxiv.org/abs/2205.11498).

In recent news, [Matryoshka Representation Learning](https://arxiv.org/abs/2205.13147) ([blogpost](https://huggingface.co/blog/matryoshka)) (MRL) as used by [OpenAI](https://openai.com/blog/new-embedding-models-and-api-updates) also allows for cheaper embeddings. With MRL, only the first `n` embedding dimensions are used. This approach has already been adopted by some open models like [nomic-ai/nomic-embed-text-v1.5](https://huggingface.co/nomic-ai/nomic-embed-text-v1.5) and [mixedbread-ai/mxbai-embed-2d-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-2d-large-v1) [^1]. 

> However, there is another new approach to achieve progress on this challenge; it does not entail dimensionality reduction, but rather a reduction in size of each of the individual values in the embedding: **Quantization**. Our experiments on quantization will show that we can maintain a large amount of performance while significantly speeding up computation and saving on memory, storage, and costs. Let's dive into it! 

Reviewer: ```suggestion
However, there is another new approach to achieve progress on this challenge; it does not entail dimensionality reduction, but rather a reduction in the size of each of the individual values in the embedding: **Quantization**. Our experiments on quantization will show that we can maintain a large amount of performance while significantly speeding up computation and saving on memory, storage, and costs. Let's dive into it! 
```

> ### Binary Quantization

Reviewer: People are familiar with model quantization. Maybe it's worth clarifying that this is different?

Author: Tackled via c8ce9a3e3f600f90b5b5e1ecb37a9c9946cedf57, feel free to review

Reviewer 2: Looks good to me!

Binary quantization refers to the conversion of the `float32` values in an embedding to 1-bit values, resulting in a 32x reduction in memory and storage usage.

To quantize `float32` embeddings to binary, we simply threshold normalized embeddings at 0:

$$
    f(x)= 
\begin{cases}
    0 & \text{if } x\leq 0\\
    1 & \text{if } x \gt 0
\end{cases}
$$

> We can use the Hamming Distance to efficiently perform retrieval with these binary embeddings. This is simply the number of positions at which the bits of two binary embeddings differ. The lower the Hamming Distance, the closer the embeddings, and thus the more relevant the document. A huge advantage of the Hamming Distance is that it can be easily calculated with 2 CPU cycles, allowing for blazingly fast performance.

Reviewer: ```suggestion
We can use the Hamming Distance to retrieve these binary embeddings efficiently. This is the number of positions at which the bits of two binary embeddings differ. The lower the Hamming Distance, the closer the embeddings; thus, the more relevant the document. A huge advantage of the Hamming Distance is that it can be easily calculated with 2 CPU cycles, allowing for blazingly fast performance.
```

#### Binary Quantization in Sentence Transformers

Quantizing an embedding with a dimensionality of 1024 to binary would result in 1024 bits. In practice, it is much more common to store bits as bytes instead, so when we quantize to binary embeddings, we pack the bits into bytes using `np.packbits`.

Therefore, quantizing a `float32` embedding with a dimensionality of 1024 yields an `int8` or `uint8` embedding with a dimensionality of 128. See two approaches of how you can produce quantized embeddings using [Sentence Transformers](https://sbert.net/) below:

```python
from sentence_transformers import SentenceTransformer

# 1. Load an embedding model
model = SentenceTransformer("mixedbread-ai/mxbai-embed-large-v1")

# 2a. Encode some text using "binary" quantization
binary_embeddings = model.encode(
    ["I am driving to the lake.", "It is a beautiful day."],
    precision="binary",
)
```
or
```python

> from sentence_transformers import SentenceTransformer

Reviewer: I would not reimport things that are imported in previous cells

Author: These are meant to be ran separately, either the first or the second. That's why I kept both snippets fully copy-pastable.

Reviewer: SG!

from sentence_transformers.quantization import quantize_embeddings

# 1. Load an embedding model
model = SentenceTransformer("mixedbread-ai/mxbai-embed-large-v1")

# 2b. or, encode some text without quantization & apply quantization afterwards
embeddings = model.encode(["I am driving to the lake.", "It is a beautiful day."])
binary_embeddings = quantize_embeddings(embeddings, precision="binary")
```

**References:**
* <a href="https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1"><code>mixedbread-ai/mxbai-embed-large-v1</code></a>
* <a href="https://sbert.net/docs/package_reference/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode"><code>SentenceTransformer.encode</code></a>
* <a href="https://sbert.net/docs/package_reference/quantization.html#sentence_transformers.quantization.quantize_embeddings"><code>quantize_embeddings</code></a>

> Here you can see the differences between default `float32` embeddings and binary embeddings in terms of shape, size, and `numpy` dtype:

Reviewer: ```suggestion
Here, you can see the differences between default `float32` embeddings and binary embeddings in terms of shape, size, and `numpy` dtype:
```

```python
>>> embeddings.shape
(2, 1024)
>>> embeddings.nbytes
8192
>>> embeddings.dtype
float32
>>> binary_embeddings.shape
(2, 128)
>>> binary_embeddings.nbytes
256
>>> binary_embeddings.dtype
int8
```
Note that you can also choose `"ubinary"` to quantize to binary using the unsigned `uint8` data format. This may be a requirement depending on your vector library/database.

#### Binary Quantization in Vector Databases

| Vector Databases | Support |
| - | -|
| Faiss | [Yes](https://github.com/facebookresearch/faiss/wiki/Binary-indexes) |
| USearch | [Yes](https://github.com/unum-cloud/usearch) |
| Vespa AI | [Yes](https://docs.vespa.ai/en/reference/schema-reference.html) |
| Milvus | [Yes](https://milvus.io/docs/index.md) |
| Qdrant | Through [Binary Quantization](https://qdrant.tech/documentation/guides/quantization/#binary-quantization) |
| Weaviate | Through [Binary Quantization](https://weaviate.io/developers/weaviate/configuration/bq-compression) |

### Scalar (int8) Quantization

> To convert the `float32` embeddings into `int8`, we use a process called scalar quantization. This involves mapping the continuous range of `float32` values to the discrete set of `int8` values, which can represent 256 distinct levels (from -128 to 127) as shown in the image below. This is done by using a large calibration dataset of embeddings. We compute the range of these embeddings, i.e. the `min` and `max` of each of the embedding dimensions. From there, we calculate the steps (buckets) in which we categorize each value.

Reviewer: ```suggestion
We use a scalar quantization process to convert the `float32` embeddings into `int8`. This involves mapping the continuous range of `float32` values to the discrete set of `int8` values, which can represent 256 distinct levels (from -128 to 127), as shown in the image below. This is done by using a large calibration dataset of embeddings. We compute the range of these embeddings, i.e., the `min` and `max` of each embedding dimension. From there, we calculate the steps (buckets) to categorize each value.
```

> To further boost the retrieval performance, you can optionally apply the same rescoring step as for the binary embeddings. It is important to note here that the calibration dataset has a large influence on performance, since it defines the quantization buckets.

Reviewer: ```suggestion
To further boost the retrieval performance, you can optionally apply the same rescoring step as for the binary embeddings. It is important to note that the calibration dataset greatly influences performance since it defines the quantization buckets.
```

<p align="center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/embedding-quantization/scalar-quantization.png">
    <em><small>Source: <a href="https://qdrant.tech/articles/scalar-quantization/">https://qdrant.tech/articles/scalar-quantization/</a></small></em>
</p>

> With scalar quantization to `int8`, we reduce the precision of the original `float32` embeddings such that each value is represented with an 8-bit integer (4x smaller). Note that this differs from the binary quantization case, where each value is represented by a single bit (32x smaller).

Reviewer: ```suggestion
With scalar quantization to `int8`, we reduce the original `float32` embeddings' precision so that each value is represented with an 8-bit integer (4x smaller). Note that this differs from the binary quantization case, where each value is represented by a single bit (32x smaller).
```

Here you can see the differences between default `float32` embeddings and `int8` scalar embeddings in terms of shape, size, and `numpy` dtype:

```python
>>> embeddings.shape
(2, 1024)
>>> embeddings.nbytes
8192
>>> embeddings.dtype
float32
>>> int8_embeddings.shape
(2, 1024)
>>> int8_embeddings.nbytes
2048
>>> int8_embeddings.dtype
int8
```

#### Scalar Quantization in Vector Databases

| Vector Databases | Support |
| - | - |
| Faiss | Indirectly through [IndexHNSWSQ](https://faiss.ai/cpp_api/struct/structfaiss_1_1IndexHNSWSQ.html) |
| USearch | [Yes](https://github.com/unum-cloud/usearch) | 
| Vespa AI |  [Yes](https://docs.vespa.ai/en/reference/tensor.html) |
| OpenSearch | [Yes](https://opensearch.org/docs/latest/field-types/supported-field-types/knn-vector) |
| ElasticSearch | [Yes](https://www.elastic.co/de/blog/save-space-with-byte-sized-vectors) |
| Milvus | Indirectly through [IVF_SQ8](https://milvus.io/docs/index.md) |
| Qdrant | Indirectly through [Scalar Quantization](https://qdrant.tech/documentation/guides/quantization/#scalar-quantization) |

### Combining Binary and Scalar Quantization

> It is possible to combine binary and scalar quantization to get the best of both worlds: the extreme speed from binary embeddings and the great performance preservation of scalar embeddings with rescoring. See the [demo](#demo) below for a real-life implementation of this approach involving 41 million texts from Wikipedia. The pipeline for that setup is as follows:

Reviewer: ```suggestion
Combining binary and scalar quantization is possible to get the best of both worlds: the extreme speed from binary embeddings and the great performance preservation of scalar embeddings with rescoring. See the [demo](#demo) below for a real-life implementation of this approach involving 41 million texts from Wikipedia. The pipeline for that setup is as follows:
```

1. The query is embedded using the [`mixedbread-ai/mxbai-embed-large-v1`](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) SentenceTransformer model.
2. The query is quantized to binary using the <a href="https://sbert.net/docs/package_reference/quantization.html#sentence_transformers.quantization.quantize_embeddings"><code>quantize_embeddings</code></a> function from the `sentence-transformers` library.
3. A binary index (41M binary embeddings; 5.2GB of memory/disk space) is searched using the quantized query for the top 40 documents.
4. The top 40 documents are loaded on the fly from an int8 index on disk (41M int8 embeddings; 0 bytes of memory, 47.5GB of disk space).
5. The top 40 documents are rescored using the float32 query and the int8 embeddings to get the top 10 documents.
6. The top 10 documents are sorted by score and displayed.

> Through this approach, we use 5.2GB of memory and 52GB of disk space for the indices. This is considerably less than normal retrieval, for which we would require 200GB of memory and 200GB of disk space. Especially as you scale up even further, this will result in notable reductions in both latency and costs.

Reviewer: ```suggestion
Through this approach, we use 5.2GB of memory and 52GB of disk space for the indices. This is considerably less than normal retrieval, requiring 200GB of memory and 200GB of disk space. Especially as you scale up even further, this will result in notable reductions in latency and costs.
```

---
title: "Binary and Scalar Embedding Quantization for Significantly Faster & Cheaper Retrieval"
thumbnail: /blog/assets/embedding-quantization/thumbnail.png
authors:
- user: angygraycat
  guest: true
- user: tomaarsen
- user: SeanLee97
  guest: true
---

# Binary and Scalar Embedding Quantization for Significantly Faster & Cheaper Retrieval

> In this blogpost, we will introduce the concept of embedding quantization and showcase their impact on retrieval speed, memory usage, disk space, and costs. We will discuss how embeddings can be quantized in theory and in practice, after which we introduce a demo showing a real-life retrieval scenario of 41 million Wikipedia texts.

Reviewer 2: ```suggestion
We introduce the concept of embedding quantization and showcase their impact on retrieval speed, memory usage, disk space, and cost. We'll discuss how embeddings can be quantized in theory and in practice, after which we introduce a demo showing a real-life retrieval scenario of 41 million Wikipedia texts.
```

> ## Table of Contents

Reviewer 2: Just FYI, there's an automatic TOC UI component like now, you have to click on the hamburger button:

<img width="446" alt="Screenshot 2024-03-22 at 14 12 16" src="https://github.com/huggingface/blog/assets/1177582/6070c7da-7917-45ad-84e2-eeaef9539fef">

Author: Would you recommend that we remove the "manual" ToC?

Reviewer 2: Your call, I think users may not see / be familiar with the automatic one yet, so I think it's ok to keep it if you think it helps, especially considering that the post has several subsections.

Author: We'll keep it

    + [Binary Quantization](#binary-quantization)
        - [Binary Quantization in Sentence Transformers](#binary-quantization-in-sentence-transformers)
        - [Binary Quantization in Vector Databases](#binary-quantization-in-vector-databases)
    + [Scalar (int8) Quantization](#scalar-int8-quantization)
        - [Scalar Quantization in Sentence Transformers](#scalar-quantization-in-sentence-transformers)
        - [Scalar Quantization in Vector Databases](#scalar-quantization-in-vector-databases)
    + [Combining Binary and Scalar Quantization](#combining-binary-and-scalar-quantization)
    + [Quantization Experiments](#quantization-experiments)
    + [Influence of Rescoring](#influence-of-rescoring)
        - [Binary Rescoring](#binary-rescoring)
        - [Scalar (Int8) Rescoring](#scalar-int8-rescoring)
        - [Retrieval Speed](#retrieval-speed)
    + [Performance Summarization](#performance-summarization)
    + [Demo](#demo)
    + [Try it yourself](#try-it-yourself)
    + [Future work:](#future-work)
    + [Acknowledgments](#acknowledgments)
    + [Citation](#citation)
    + [References](#references)

## Why Embeddings?

> Embeddings are one of the most versatile tools in natural language processing, supporting their users in a variety of different settings and use cases. In essence, an embedding is a numerical representation of more complex objects like, for example, text, images, or audio. Specifically, the objects are represented as n-dimensional vectors. 

Reviewer 2: ```suggestion
Embeddings are one of the most versatile tools in natural language processing, supporting a wide variety of settings and use cases. In essence, embeddings are numerical representations of more complex objects, like text, images, audio, etc. Specifically, the objects are represented as n-dimensional vectors. 
```

After transforming the complex objects, you can then determine their similarity by calculating the similarity of the respective embeddings! This is crucial for a massive amount of use cases: it serves as the backbone for recommendation systems, retrieval, one-shot or few-shot learning, outlier detection, similarity search, paraphrase detection, clustering, classification, and much more.

### Embeddings may struggle to scale

> However, these embeddings may be challenging to scale up, which leads to expensive solutions and high latencies. Currently, many state-of-the-art models produce embeddings with 1024 dimensions, each of which is encoded in `float32`, i.e., they require 4 bytes per dimension. To perform retrieval over 250 million vectors, you would therefore need around 1TB of memory!

Reviewer 2: ```suggestion
However, embeddings may be challenging to scale for production use cases, which leads to expensive solutions and high latencies. Currently, many state-of-the-art models produce embeddings with 1024 dimensions, each of which is encoded in `float32`, i.e., they require 4 bytes per dimension. To perform retrieval over 250 million vectors, you would therefore need around 1TB of memory!
```

The table below gives an overview of different models, dimension size, memory requirement, and costs. Costs are computed at an estimated $3.8 per GB/mo with `x2gd` instances on AWS.

> <!--For example, the model [mxbai-embed-large-v1]() from mixedbread ai produces vectors which have 1024 dimensions. Every dimension is encoded in float32 - so 4 bytes per dimension. To store 250 Million vectors you will need around 1TB  of memory! The table below shows examples of different models, dimension, memory requirement and costs.-->
> 
> 
> <Table with costs for 1M, 10M, 100M, 1B embeddings at ?>
> 

Reviewer 2: ```suggestion
```

Author: Whoops!

| 1024                | [bge-large-en-v1.5](https://huggingface.co/BAAI/bge-large-en-v1.5)<br>[mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1)<br>[Cohere-embed-english-v3.0](https://txt.cohere.com/introducing-embed-v3/)                                                                                                                                                        |381.46GB<br>$1,449 / mo|953.67GB<br>$3,623 / mo|3814.69GB<br>$14,495 / mo|
| 1536                | [OpenAI text-embedding-3-small](https://openai.com/blog/new-embedding-models-and-api-updates) |572.20GB<br>$2,174 / mo|1430.51GB<br>$5,435 / mo|5722.04GB<br>$21,743 / mo|
| 3072                | [OpenAI text-embedding-3-large](https://openai.com/blog/new-embedding-models-and-api-updates) |1144.40GB<br>$4,348 / mo|2861.02GB<br>$10,871 / mo|11444.09GB<br>$43,487 / mo|

> ## Improving scalability
> They are several ways to approach the challenges of scaling embeddings. The most common approach is dimensionality reduction, such as PCA [^1]. However, classic dimensionality reduction -- like PCA methods -- tends to perform poorly when used with embeddings [^5].

Reviewer 2: ```suggestion
## Improving scalability

There are several ways to approach the challenges of scaling embeddings. The most common approach is dimensionality reduction, such as PCA [^1]. However, classic dimensionality reduction -- like PCA methods -- tends to perform poorly when used with embeddings [^5].
```

> In recent news, Matryoshka Representation Learning [^2][^3] (MRL) as used by OpenAI [^4] also allows for cheaper embeddings. With MRL, only the first `n` embedding dimensions are used. This approach has already been adopted by some open models like [nomic-ai/nomic-embed-text-v1.5](https://huggingface.co/nomic-ai/nomic-embed-text-v1.5) and [mixedbread-ai/mxbai-embed-2d-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-2d-large-v1) [^6]. 

Reviewer 2: ```suggestion
In recent news, Matryoshka Representation Learning [^2][^3] (MRL), as used by OpenAI [^4], also allows for cheaper embeddings. With MRL, only the first `n` embedding dimensions are used. This approach has already been adopted by some open models like [nomic-ai/nomic-embed-text-v1.5](https://huggingface.co/nomic-ai/nomic-embed-text-v1.5) and [mixedbread-ai/mxbai-embed-2d-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-2d-large-v1) [^6]. 
```

Why don't we use url links instead? I find them less distracting, less academic looking (so more welcoming to regular users), and more natural in a blog post.

Author: cc @aamir-s18 thoughts?

Reviewer 3: lets do links

Author: I'll make the updates

Author: Done in ed3b3f448889f27d5124ee5f7ae2fe7fd44c5e0e

> However, there is another new approach to achieve progress on this challenge; it entails no reduction of the dimensionality of the embeddings, but rather a reduction in size of each of the individual values in the embedding: **Quantization** [^7][^8][^9]. Our experiments on quantization will show that we can maintain a large amount of performance while significantly speeding up computation and saving on memory, storage, and costs. Let's dive into it! 

Reviewer 2: ```suggestion
However, there is another new approach to achieve progress on this challenge; it entails dimensionality reduction, but rather a reduction in size of each of the individual values in the embedding: **Quantization** [^7][^8][^9]. Our experiments on quantization will show that we can maintain a large amount of performance while significantly speeding up computation and saving on memory, storage, and costs. Let's dive into it! 
```

<!--
With quantization, we change the amount of information used for every entry in vector, so that embedding takes less space. This is done by exchanging precision (bits per entry) with speed and storage cost. Our experiment in the following will show that we cen maintain a large amount of performance, gain significant speedups and save storage. Let's dive into it. 
-->

### Binary Quantization

> <!--Binary embeddings convert the float32 values to 1 bit values, giving us a 32x reduction in memory. -->

Reviewer 2: I'd remove all the comments before publishing

Author: Will do

Author: Done in 4b791c0525874fd0d25b4ef24c87b01b31bbdcda

Binary quantization refers to the conversion of the `float32` values in an embedding to 1-bit values, resulting in a 32x reduction in memory and storage usage.

<!--To quantize to binary, we apply an approach inspired by Yamada et al. (2021) [^7].-->

<!--A classic approach for binary quantization is to apply hashing of some kind, like locality-sensitive hashing. Here we take a hashing from float embeddings to binary embeddings. However as it was shown in "Efficient Passage Retrieval with Hashing for Open-domain Question Answering"[^7] those methods works not so well. Instead take the approach introduced in this work called Binary Passage Retrieval. -->

To quantize `float32` embeddings to binary, we simply threshold normalized embeddings at 0:<!-- if the value is smaller than 0, we make it 0 else 1. -->

$$
    f(x)= 
\begin{cases}
    0 & \text{if } x\leq 0\\
    1 & \text{if } x \gt 0
\end{cases}
$$

> We can use the Hamming Distance to efficiently perform retrieval with these binary embeddings. This is simply the number of positions at which the bits of two binary embeddings differ. The lower the Hamming Distance, the closer the embeddings, and thus the more relevant the document. A huge advantage of the Hamming Distance is that it can be easily calculated with 2 CPU cycles, allowing for blazingly fast performance.

Reviewer 2: 🔥

> Yamada et al. (2021)[^10] introduced a rescore step, which they called *rerank*, to boost the performance. They proposed that the `float32` query embedding could be compared with the binary document embeddings using dot-product. In practice, we first retrieve `rescore_multiplier * top_k` results with the binary query embedding and the binary document embeddings -- i.e., the list of the first k results of the double-binary retrieval --  and then rescore that list of binary document embeddings with the `float32` query embedding.

Reviewer 2: Just as a suggestion, I'd maybe try to write this in a more casual fashion so it feels less like a paper. We can still cite or link to Yamada's work, of course.

> By applying this novel rescoring step, we are able to preserve up to ~96% of the total retrieval performance, while reducing the memory and disk space usage by 32x and improving the retrieval speed by up to 32x as well.

Reviewer 2: Nice! Do we have numbers about the performance quality without the reranking step? It could be interesting to show the difference :)

Reviewer 3: They are further below, in the section about influence of rescoring.

Reviewer 3: @tomaarsen

#### Binary Quantization in Sentence Transformers

Quantizing an embedding with a dimensionality of 1024 to binary would result in 1024 bits. In practice, it is much more common to store bits as bytes instead, so when we quantize to binary embeddings, we pack the bits into bytes using `np.packbits`.

> As a result, in practice quantizing a `float32` embedding with a dimensionality of 1024 yields an `int8` or `uint8` embedding with a dimensionality of 128. See two approaches of how you can produce quantized embeddings using [Sentence Transformers](https://sbert.net/) below:

Reviewer 2: ```suggestion
Therefore, quantizing a `float32` embedding with a dimensionality of 1024 yields an `int8` or `uint8` embedding with a dimensionality of 128. See two approaches of how you can produce quantized embeddings using [Sentence Transformers](https://sbert.net/) below:
```

```python
from sentence_transformers import SentenceTransformer
from sentence_transformers.quantization import quantize_embeddings

# 1. Load an embedding model
model = SentenceTransformer("mixedbread-ai/mxbai-embed-large-v1")

# 2a. Encode some text using "binary" quantization
binary_embeddings = model.encode(
    ["I am driving to the lake.", "It is a beautiful day."],
    precision="binary",
)

# 2b. or, encode some text without quantization & apply quantization afterwards
embeddings = model.encode(["I am driving to the lake.", "It is a beautiful day."])
binary_embeddings = quantize_embeddings(embeddings, precision="binary")

> ```

Reviewer 2: Maybe clearer if we separate the code blocks? Not sure, but I feel some readers may not read comments at first.

Author: agreed, done in af2165644ae6d0c463aafade096529afe509c661

**References:**
* <a href="https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1"><code>mixedbread-ai/mxbai-embed-large-v1</code></a>
* <a href="https://sbert.net/docs/package_reference/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode"><code>SentenceTransformer.encode</code></a>

> * <a href="https://sbert.net/docs/package_reference/quantization.html#sentence_transformers.quantization.quantize_embeddings"><code>quantize_embeddings</code></a>

Reviewer 2: Maybe also copy to a "Resources" section at the end of the post, with everything together.

Author: Done!

Here you can see the differences between default `float32` embeddings and binary embeddings in terms of shape, size, and `numpy` dtype:

```python
>>> embeddings.shape
(2, 1024)
>>> embeddings.nbytes
8192
>>> embeddings.dtype
float32
>>> binary_embeddings.shape
(2, 128)
>>> binary_embeddings.nbytes
256
>>> binary_embeddings.dtype
int8
```

> Note that you can also choose `"ubinary"` to quantize to binary using the unsigned `uint8` data format. This may be a requirement for your vector library/database.

Reviewer 2: ```suggestion
Note that you can also choose `"ubinary"` to quantize to binary using the unsigned `uint8` data format. This may be a requirement depending on your vector library/database.
```

#### Binary Quantization in Vector Databases

| Vector Databases | Support |
| - | -|
| Faiss | [Yes](https://github.com/facebookresearch/faiss/wiki/Binary-indexes) |
| USearch | [Yes](https://github.com/unum-cloud/usearch) |
| Vespa AI | [Yes](https://docs.vespa.ai/en/reference/schema-reference.html) |
| Milvus | [Yes](https://milvus.io/docs/index.md) |
| Qdrant | Through [Binary Quantization](https://qdrant.tech/documentation/guides/quantization/#binary-quantization) |
| Weaviate | Through [Binary Quantization](https://weaviate.io/developers/weaviate/configuration/bq-compression) |

### Scalar (int8) Quantization

> To convert the `float32` embeddings into `int8`, we use a process called scalar quantization. This involves mapping the continuous range of `float32` values to the discrete set of `int8` values, which can represent 256 distinct levels (from -128 to 127) as shown in the image below. This is done by using a large calibration dataset of embeddings. We compute the range of these embeddings, i.e. the `min` and `max` of each of the embedding dimensions. From there, we calculate the steps (buckets) in which we categorize each value.

Reviewer 2: Maybe clarify that this is different than binary quantization packed as 8-bit integers, to avoid any confusion about the previous section also producing `int8` values.

For example (just a suggestion):

> With scalar quantization to `int8`, we reduce the precision of the original `float32` embeddings so each value is represented with an 8-bit integer. Note that in the binary quantization case, each value is represented by a single bit.

Author: Agreed, addressed in c2eaef44aafbd3545d9af3046f984287cefd16b7

> To further boost the retrieval performance, you can optionally apply the same rescoring step as for the binary embeddings. It is important to note here that the calibration dataset has a large influence on the performance, since it defines the buckets.

Reviewer 2: ```suggestion
To further boost the retrieval performance, you can optionally apply the same rescoring step as for the binary embeddings. It is important to note here that the calibration dataset has a large influence on performance, since it defines the quantization buckets.
```

Is there any recommendation or comment about the homogeneity of the embeddings so we prevent "out of context" embeddings for inference?

Author: Not from my side, perhaps @aamir-s18 has a comment to add here

Reviewer 3: Using a large dataset like wikipedia is fine. The larger the dataset the less the influence and the more stable the results across all benchmarks.

<p align="center">

>     <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/embedding-quantization/scalar-quantization.png">

Reviewer 2: Can't see the image, will check if there's an open PR.

Reviewer 2: It works now

If none of these are the case, you will be given a warning like this:
`Computing int8 quantization buckets based on 2 embeddings. int8 quantization is more stable with 'ranges' calculated from more embeddings or a 'calibration_embeddings' that can be used to calculate the buckets.`

See how you can produce scalar quantized embeddings using [Sentence Transformers](https://sbert.net/) below:

```python
from sentence_transformers import SentenceTransformer
from sentence_transformers.quantization import quantize_embeddings
from datasets import load_dataset

# 1. Load an embedding model
model = SentenceTransformer("mixedbread-ai/mxbai-embed-large-v1")

# 2. Prepare an example calibration dataset
corpus = load_dataset("nq_open", split="train[:1000]")["question"]
calibration_embeddings = model.encode(corpus)

# 3. Encode some text without quantization & apply quantization afterwards
embeddings = model.encode(["I am driving to the lake.", "It is a beautiful day."])
int8_embeddings = quantize_embeddings(
    embeddings,
    precision="int8",
    calibration_embeddings=calibration_embeddings,

> )

Reviewer 2: Interesting. Can the calibration step be cached for subsequent use with different embeddings?

Author: Yes, in practice you can also provide `ranges`, which is a much smaller array (`[2, num_dimensions]`), which should be used in that case.

1. The query is embedded using the [`mixedbread-ai/mxbai-embed-large-v1`](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) SentenceTransformer model.
2. The query is quantized to binary using the <a href="https://sbert.net/docs/package_reference/quantization.html#sentence_transformers.quantization.quantize_embeddings"><code>quantize_embeddings</code></a> function from the `sentence-transformers` library.
3. A binary index (41M binary embeddings; 5.2GB of memory/disk space) is searched using the quantized query for the top 40 documents.
4. The top 40 documents are loaded on the fly from an int8 index on disk (41M int8 embeddings; 0 bytes of memory, 47.5GB of disk space).
5. The top 40 documents are rescored using the float32 query and the int8 embeddings to get the top 10 documents.
6. The top 10 documents are sorted by score and displayed.

> Through this approach, we use 130MB of memory and 1.3GB of disk space for the indices. This is considerably less than normal retrieval, for which we would require 4GB of memory and 4GB of disk space. Especially as you scale up even further, this will result in notable reductions in both latency and costs.

Reviewer 2: Very cool! Do we have the final quality performance of this method?

Reviewer 3: Unfortunately not yet, we are on it, but it will take some time. We add them in a edit?

Reviewer 2: Sure!

The benefits of quantization are, if anything, even more clearly visible when looking at the results obtained with binary models. In that scenario, the 1024 dimension models still outperform a now 10x more storage intensive base model, and the `mxbai-embed-large-v1` even manages to hold more than 96% of performance after a 32x reduction in resource requirements. The further quantization from `int8` to binary barely results in any additional loss of performance for this model. 

Interestingly, we can also see that `all-MiniLM-L6-v2` exhibits stronger performance on binary than on `int8` quantization. A possible explanation for this could be the selection of calibration data. On `e5-base-v2`, we observe the effect of dimension collapse[^12], which causes the model to only use a subspace of the latent space; when performing the quantization, the whole space collapses further, leading to high performance losses.

This shows that quantization doesn't universally work with all embedding models. It remains crucial to consider exisiting benchmark outcomes and conduct experiments to determine a given model's compatibility with quantization.

### Influence of Rescoring

> In this chapter we will investigate the influence of rescoring on the retrieval performance. We are evaluating the results based on [mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1).

Reviewer 2: ```suggestion
In this section we look at the influence of rescoring on retrieval performance. We evaluate the results based on [mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1).
```

#### Binary Rescoring

> For binary embeddings [mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) retains 92.53% of performance on MTEB Retrieval. Just doing the rescoring without retrieving more samples pushes the performance to 96.45%. We experimented with setting the`rescore_multiplier` from 1 to 10, but observe no further boost in performance. This indicates that the `top_k` search already retrieved the top candidates and the rescoring reordered these good candidates appropriately.

Reviewer 2: ```suggestion
With binary embeddings, [mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) retains 92.53% of performance on MTEB Retrieval. Just doing the rescoring without retrieving more samples pushes the performance to 96.45%. We experimented with setting the`rescore_multiplier` from 1 to 10, but observe no further boost in performance. This indicates that the `top_k` search already retrieved the top candidates and the rescoring reordered these good candidates appropriately.
```

🔥

#### Scalar (Int8) Rescoring

> We additionally evaluated the [mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) model with `int8` rescoring, as Cohere showed that [Cohere-embed-english-v3.0](https://txt.cohere.com/introducing-embed-v3/) reached up to 100% of the performance of the `float32` model with `int8` quantization. For this experiment, we set the `rescore_multiplier` to [1, 4, 10] and received the following results:

Reviewer 2: ```suggestion
We also evaluated the [mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) model with `int8` rescoring, as Cohere showed that [Cohere-embed-english-v3.0](https://txt.cohere.com/introducing-embed-v3/) reached up to 100% of the performance of the `float32` model with `int8` quantization. For this experiment, we set the `rescore_multiplier` to [1, 4, 10] and got the following results:
```

<p align="center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/embedding-quantization/rescoring-influence.png">
</p>

As we can see from the diagram, a higher rescore multiplier implies better retention of performance after quantization. Extrapolating from our results, we assume the relation is likely hyperbolical with performance approaching 100% as the rescore multiplier continues to rise. A rescore multiplier of 4-5 already leads to a remarkable performance retention of 99% using `int8`.

#### Retrieval Speed

> The retrieval speed was measures on Google Cloud Platform `a2-highgpu-4g` Instance using the [mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) embeddings with 1024 dimension on the whole MTEB Retrieval. For int8 we used [USearch](https://github.com/unum-cloud/usearch) (Version 2.9.2) and binary quantization [Faiss](https://github.com/facebookresearch/faiss) (Version 1.8.0). Everything was computed on CPU using exact search.

Reviewer 2: ```suggestion
We measured retrieval speed on a Google Cloud Platform `a2-highgpu-4g` instance using the [mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) embeddings with 1024 dimension on the whole MTEB Retrieval. For int8 we used [USearch](https://github.com/unum-cloud/usearch) (Version 2.9.2) and binary quantization [Faiss](https://github.com/facebookresearch/faiss) (Version 1.8.0). Everything was computed on CPU using exact search.
```

Author: I think because these instances were also used to compute the full float32 embeddings originally, which is notably faster on GPU.

Reviewer 3: We used them because, we created embeddings on it and it is my little cluster I am using.

| Quantization | Min            | Mean               | Max           |
|--------------|----------------|--------------------|---------------|
| `float32`    | 1x (baseline)  | **1x** (baseline)  | 1x (baseline) |
| `int8`       | 2.99x speedup  | **3.66x** speedup  | 4.8x speedup  |
| `binary`     | 15.05x speedup | **24.76x** speedup | 45.8x speedup |

> As indicated by the above table, applying `int8` scalar quantization resulted in a mean speedup of 3.66x compared to full-size `float32` embeddings. Additionally, binary quantization resulted in a mean speedup of 24.76x. For both scalar and binary quantization, even the worst case scenario resulted in very notable speedups.

Reviewer 2: ```suggestion
As shown in the table, applying `int8` scalar quantization results in an average speedup of 3.66x compared to full-size `float32` embeddings. Additionally, binary quantization achieves a speedup of 24.76x on average. For both scalar and binary quantization, even the worst case scenario resulted in very notable speedups.
```

<!-- We have to check if this table make sense or not! Perhaps we can merge the theoretical 32x and 4x speedup with the previous section of "Retrieval Speed" and then remove this section -->

The experimental results, effects on resource use, retrieval speed, and retrieval performance by using quantization can be summarized as follows:

|                               | float32 | int8/uint8 | binary/ubinary |
|-------------------------------|---------:|------------:|----------------:|
| **Memory & Index size savings** | 1x      | exactly 4x         | exactly 32x            |
| **Retrieval Speed**           | 1x      | up to 4x   | up to 45x     |
| **Percentage of default performance**         | 100%    | ~99.3%     | ~96%           |

### Demo

The following demo showcases the retrieval efficiency using `exact` search through combining binary search with scalar (`int8`) rescoring. The solution requires 5GB of memory for the binary index and 50GB of disk space for the binary and scalar indices, considerably less than the 200GB of memory and disk space which would be required for regular `float32` retrieval. Additionally, retrieval is much faster.

<iframe

> 	src="https://sentence-transformers-quantized-retrieval.hf.space"

Reviewer 2: I get a 404 here

Author: Indeed, we are still working on this. The Space is being a bit difficult due to its large size.

Author: Should work now

	frameborder="0"
	width="100%"
	height="1000"
></iframe>

### Try it yourself

The following scripts can be used to experiment with embedding quantization for retrieval & beyond. There are three categories:

* **Recommended Retrieval**:

>   * [semantic_search_recommended.py](semantic_search_recommended.py): This script combines binary search with scalar rescoring, much like the above demo, for cheap, efficient, and performant retrieval.

Reviewer 2: These links seem incomplete

Author: Fixed in 36a71b0ec4799291b5b6491227ea6de7389b0cdd, thanks

  * [semantic_search_usearch_benchmark.py](semantic_search_usearch_benchmark.py): This script includes a retrieval speed benchmark of `float32` retrieval, binary retrieval + rescoring, and scalar retrieval + rescoring, using USearch. It uses the <a href="https://sbert.net/docs/package_reference/quantization.html#sentence_transformers.quantization.semantic_search_usearch"><code>semantic_search_usearch</code></a> utility function. Our experiments show large speedups on newer hardware, particularly for `int8`.

### Future work

We are looking forward to further advancements of binary quantization. To name a few potential improvements, we suspect that there may be room for scalar quantization smaller than `int8`, i.e. with 128 or 64 buckets instead of 256.

> Additionally, we are excited that embedding quantization is fully perpendicular to Matryoshka Representation Learning (MRL). With other words, it is possible to shrink MRL embeddings from e.g. 1024 to 128 (which usually corresponds with a 2% reduction in performance) and then apply binary or scalar quantization. We suspect this could speed up retrieval up to 32x for a ~3% reduction in quality or up to 256x for a ~10% reduction in quality.

Reviewer 2: ```suggestion
Additionally, we are excited that embedding quantization is fully perpendicular to Matryoshka Representation Learning (MRL). In other words, it is possible to shrink MRL embeddings from e.g. 1024 to 128 (which usually corresponds with a 2% reduction in performance) and then apply binary or scalar quantization. We suspect this could speed up retrieval up to 32x for a ~3% reduction in quality, or up to 256x for a ~10% reduction in quality.
```

### Binary Quantization

> Unlike quantization in models where you reduce the precision of weights, quantization for embeddings refers to a post-processing step for the embeddings themselves. In particular, binary quantization refers to the conversion of the `float32` values in an embedding to 1-bit values, resulting in a 32x reduction in memory and storage usage.

Reviewer 2: 👍
